shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -182,6 +183,115 @@
+//
+// Clips are applied by adding a super view with a CALayer mask. Transforms are applied to the
+// view that's at the head of the chain. For example the following mutators stack [T_1, C_2, T_3,
+// T_4, C_5, T_6] where T denotes a transform and C denotes a clip, will result in the following
+// UIView tree:
+//
+// C_2 -> C_5 -> PLATFORM_VIEW
+// (PLATFORM_VIEW is a subview of C_5 which is a subview of C_2)
+//
+// T_1 is applied to C_2, T_3 and T_4 are applied to C_5, and T_6 is applied to PLATFORM_VIEW.
+//
+// UIView instances used for clip mutations are re-used when possible.
+// For example if the mutators stack above is updated to [T_1, C_7, T_3, T_4, C_8, T_9], we do not
+// remove the previous superviews and re-create new views. Instead, we reuse the superviews that
+// were used to perform clips in the previous example and perform the new clips on the same views.
+//
+// Returns the UIView that's at the top of the generated chain(C_2 for the example above).
+UIView* FlutterPlatformViewsController::ApplyMutators(const MutatorsStack& mutators_stack,
+                                                      UIView* embedded_view,
+                                                      int view_id) {
+  UIView* head = embedded_view;
+  head.clipsToBounds = YES;
+
+  // Reset everything related to transform.
+  head.layer.transform = CATransform3DIdentity;
+  ResetAnchor(head.layer);
+
+  // Apply transforms/clips.
+  std::vector<std::shared_ptr<Mutator>>::const_reverse_iterator iter = mutators_stack.bottom();
+  int64_t clipCount = 0;
+  while (iter != mutators_stack.top()) {
+    switch ((*iter)->type()) {
+      case transform: {
+        CATransform3D transform = GetCATransform3DFromSkMatrix((*iter)->matrix());
+        head.layer.transform = CATransform3DConcat(head.layer.transform, transform);
+        break;
+      }
+      case clip_rect:
+      case clip_rrect:
+      case clip_path: {
+        UIView* clipView = head.superview;
+        ++clipCount;
+        if (clipCount > clip_count_[view_id]) {
+          // if we need more clip operations than last time, create a new view.
+          [head removeFromSuperview];
+          clipView = [[ChildClippingView alloc] initWithFrame:flutter_view_.get().bounds];
+          [clipView addSubview:head];
+        }
+        clipView.layer.transform = CATransform3DIdentity;
+        [(ChildClippingView*)clipView setClip:(*iter)->type()
+                                         rect:(*iter)->rect()
+                                        rrect:(*iter)->rrect()
+                                         path:(*iter)->path()];
+        ResetAnchor(clipView.layer);
+        head = clipView;
+        break;
+      }
+    }
+    ++iter;
+  }
+
+  // If we have less cilp operations this time, remove unnecessary views.
+  // We skip this process if we have more clip operations this time.
+  bool hasExtraClipViews = clip_count_[view_id] - clipCount > 0;
```

Comment:
f2f discussion: I was confused by this, lets go with the new names you suggested(`previousClipCount` `currentClipCount`)